### PR TITLE
Speed up tests ~2x

### PR DESCRIPTION
**Pull request by adamchainz**
## Description of the Change

Speed up the test suite to be about twice as fast, using these techniques:

1. Move model creation into `setUpTestData()`, so it runs only once per test case. See [my post](https://adamj.eu/tech/2021/04/12/how-to-convert-a-testcase-from-setup-to-setuptestdata/).
2. Remove `tearDown()` methods that removed model instances. These were always unnecessary as Django resets the database for you.
3. Move immutable `RequestFactory`s to class-level variables.

Measuring on the Python 3.12 and Django 5.0 environment (added in #1350), I got 207s before these changes and 105s afterwards, approximately twice as fast.

## Checklist

- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [x] documentation updated
- [x] `CHANGELOG.md` updated (only for user relevant changes)
- [x] author name in `AUTHORS`

### CODE REVIEW

**Comment by adamchainz** [comment]
`setUpTestData` copying was only added in Django 3.2, before that the [djagno-testdata package](https://github.com/charettes/django-testdata) is needed. I think we should merge #1350 first to drop Django 2.2 support, then I'll rebase this.

**Review by dopry — APPROVED** [review]
@adamchainz As a maintainer who updates a lot of branches and spends a lot of time waiting on CI spinners, I can't tell you how much I appreciate this!

**Comment by n2ygk** [comment]
@adamchainz @dopry not sure why but I've noticed that all the tests run and then the last two `success` jobs seem to hang for a long time waiting for a runner. Maybe because `success` is a different job from `build`? Is there a way to add the `Success` step into the build job somehow?  I'm not sure how to skip the matrix part within the `build` job...

**Comment by dopry** [comment]
@n2ygk, Success needs to wait on the matrix build to complete.  It can't be bundled into the same. can you open a new issue linking to an example?

**Comment by n2ygk** [comment]
@dopry not point opening a new issue if we are stuck with the current situation. Still better than the old way.

**Comment by dopry** [comment]
If you can't point to an example of the issue I can't troubleshoot it further and I suspect it is not directly related to reducing the overheat to setup test data, but is more likely related to capacity or organizational throttling by GH.  Correlation is not causation.

**Comment by n2ygk** [comment]
@dopry https://github.com/jazzband/django-oauth-toolkit/issues/1376

**Comment by n2ygk** [comment]
But I think I should have been commenting on #1219

**Pull request merged**
